### PR TITLE
Delete service ports from datapath if they are removed with a k8s update

### LINKED ISSUE #9576: Cilium keeps modified service ports in datapath
Deploy Deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: app1
spec:
  selector:
    matchLabels:
      id: app1
      zgroup: testapp
  replicas: 2
  template:
    metadata:
      labels:
        id: app1
        zgroup: testapp
    spec:
      containers:
      - name: web
        image: docker.io/cilium/demo-httpd:latest
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 80
          name: "port-80"
        - containerPort: 81
          name: "port-81"
        readinessProbe:
          httpGet:
            path: /
            port: 80
      nodeSelector:
        "cilium.io/ci-node": k8s1
```
Deploy service
```yaml
apiVersion: v1
kind: Service
metadata:
  name: app1-service
spec:
  ports:
  - name: "port-80"
    port: 80
  - name: "port-81"
    port: 80
  selector:
    id: app1
```
Verify contents of service list, it should exist as expected:
```
$ cilium service list
ID   Frontend            Service Type   Backend
...
12   172.20.0.73:81      ClusterIP                                
13   172.20.0.73:80      ClusterIP 
```
Redeploy the same service without one of the ports
```yaml
apiVersion: v1
kind: Service
metadata:
  name: app1-service
spec:
  ports:
  - name: "port-81"
    port: 81
  selector:
    id: app1
```
Verify the content of `cilium service list` is still the same. It shouldn't since port `80` is not mapping to anything anymore.
```
$ cilium service list
ID   Frontend            Service Type   Backend
...
12   172.20.0.73:81      ClusterIP                                
13   172.20.0.73:80      ClusterIP 
```

Affected versions: 
- master (def2c9e06e91d330c3dfdfc9384e65c2ad39fccf)
- v1.4.0-v1.4.9
- v1.5.0-v1.5.9
- v1.6.0-v1.6.3

### PR DESCRIPTION
**Pull request by aanm**
Fixes #9576 

Read per commit.

Note to backporters: backport the changes in `pkg/k8s/watchers/watcher.go` and `pkg/k8s/service_cache.go` only.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9577)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

File: pkg/k8s/watchers/watcher.go [hunk 1]
```diff
@@ -586,6 +568,59 @@ func (k *K8sWatcher) addK8sSVCs(svcID k8s.ServiceID, svc *k8s.Service, endpoints
...
+	svcMap := hashSVCMap(svcs)
+
+	if oldSvc != nil {
+		// If we have oldService then we need to detect which frontends
+		// are no longer in the updated service and delete them in the datapath.
+
+		oldSVCs := datapathSVCs(oldSvc, endpoints)
+		oldSVCMap := hashSVCMap(oldSVCs)
+
+		for svcHash, oldSvc := range oldSVCMap {
```
Comment (borkmann) [inline pkg/k8s/watchers/watcher.go:611]: How does this behave in terms of nodeport services where we have 3 frontend entries per service which k8s may not be aware about, would we remove e.g. the wildcard entry before re-adding them and thus create a race in the datapath?

**Comment by aanm** [comment]
test-me-please

File: pkg/k8s/watchers/watcher.go (hunk 1 above)
Comment (borkmann) [reply pkg/k8s/watchers/watcher.go:611]: Discussed offline, all good to proceed.

**Review by borkmann — APPROVED** [review]
Great catch!

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by nebril** [comment]
seems like we also need https://github.com/cilium/cilium/commit/9d64dc4d4841cf2775e59715bfa8b1a0ea88f873#diff-e5925adbfda9cd43072e9c5a10d00149 to backport this

**Pull request merged**